README.md

```diff
@@ -193,7 +193,6 @@ import (
 	"fmt"
 	"os"
 
-	homedir "github.com/mitchellh/go-homedir"
```

Comment:
@umarcor removing this will force older versions of Go to stop working. Let's not do that in this PR.

Reply:
@jharshman, I thought about that. If that was the only breaking change, I would agree. However, there is another breaking change that requires >1.11. IIRC `io.StringWriter` is not supported in deprecated versions of the language.

Reply:
I would say that my comment also applies to that change as well.
I think there is potential for a large amount of fallout from depreciating Go versions. @wfernandes opinion?

Reply:
@jharshman, note that I used `io.StringWriter` because golangci-lint would fail otherwise. In order to split it, the rule would need to be skipped per line, or per file, or the linter would need to be disabled.

I think it is better to keep this PR on hold until cobra is ready to deprecate older versions of go. Or to start a `v2` branch.

Reply:
Hmm...this is definitely a hard one. 
Ideally, I'd be all for `cobra` to declare compatible versions of go. Maybe we can revise our support for unsupported golang versions in future major releases.
However, for now it wouldn't make sense for us to break users who may be using older versions of golang and have already upgraded to v1.x.x.

I also like the use of the interface rather than passing the `*bytes.Buffer` so maybe in the interim we can replicate the interface to be something like this.
```golang
// Using a duplicate of the io.StringWriter interface so that we can support golang version <1.11
type StringWriter interface{
  WriteString(s string) (n int, err error)
}
```
The above interface type can be located at the `cobra` package level. 

Then in the future, if we ever do decide to stop supporting older versions of golang, we can simply switch out the references from `cobra.StringWriter` to `io.StringWriter`. Or if there are public methods that somehow depend on the `cobra.StringWriter` we can also do something like 
```golang
type StringWriter interface{
  io.StringWriter
}
```
This is just a suggestion. WDYT?

Reply:
Regarding the homedir stuff, maybe we just revert to the original version since it feels like it would be more work trying to work around it. 🤷

Reply:
All of the projects in https://github.com/spf13/cobra/blob/master/projects_using_cobra.md (except [delve](https://github.com/go-delve/delve)) are using >=1.12, most of them use 1.13 or 1.14 and several do require >=1.12. Honestly, I don't think that many actively maintained projects would be affected by this. Is it possible to get the +20k items in https://pkg.go.dev/github.com/spf13/cobra?tab=importedby through some API?

@wfernandes, while I think that such a solution would work now, I don't think it is acceptable in the mid-long term. I believe that having all sorts of intermediate functions to handle different versions of the language, or conditionally compiling files are acceptable solutions if required to support non EOL versions. Otherwise, it adds a severe maintainability burden and it makes it difficult to cleanup/refactor in the future. If cobra is going to support versions of golang longer than Google supports the language, I'd like to have a limit defined. For example, Debian Buster ships [1.11](https://packages.debian.org/buster/golang), so I think it would make sense to support >=1.11 but not <=1.10. By defining a limit, we can add deprecation warnings/messages for users to be aware.

Reply:
> Regarding the homedir stuff, maybe we just revert to the original version since it feels like it would be more work trying to work around it. 🤷

It is 300 loc (including sources and tests) and it has not been updated in 18 months. We might well copy it into cobra's codebase and remove the dependency.

Reply:
> or the linter would need to be disabled.

This seems like the answer here. Which linter in particular is flagging this?

Reply:
@umarcor Apologies. Didn't see your reply before I had posted mine. 
@jharshman I think that would be the `interfacer`. https://github.com/mvdan/interfacer

Reply:
I have no problem keeping the dependency a dependency.
@umarcor I think the best path forward here is to disable the linter and revert the breaking changes.

When we do eventually drop support for older Go versions, we will have to give it a bit more thought and ensure that our user impact is minimal and the change doesn't come as a surprise.

Reply:
While dropping support for older versions of Go may not be the ideal path here, I don't believe that it should be the responsibility of cobra to be the versioning "bottle neck". Any Go "1" project should in theory be compatible with all future versions of Go 1. Per the Go compatibility guarantee:
> Go programs that work today should continue to work even as future "point" releases of Go 1 arise (Go 1.1, Go 1.2, etc.). 

[Go 1 and the Future of Go Programs](https://golang.org/doc/go1compat)

Although this guarantee really only includes the standard Go API packages, cobra users who are still on go 1.11 should _in theory_ have an easy upgrade path.

In the case that cobra authors are not able to upgrade their go version, cobra 1.0.0 is feature complete and currently backwards compatible with earlier versions of go.

So my question is; why not include these? We can upgrade the version of go that cobra uses (getting all the security and performance fixes) and if users need a version of cobra that works with older versions of go, they can use the v1.0.0 tag. I do not think it should be our mandate to keep cobra (and all it's future features) compatible with all user scenarios and implementations. Rather, we should offer an upgrade path, noting the changes we make to the cobra API that are no longer able to be used with older versions of Go. Further, I would question why projects are being maintained with a version of Go that is technically EOL. There are plenty of security and performance fixes that justify cobra and it's users keeping up with the stable Go release channel.

I also recognize all the challenges this would bring. Having multiple "branches" of supported versions and compatibility creates release challenges all their own. We would likely also need to mandate a "semver" contract that would indicate how long versions are "supported".

Reply:
You have good points @jpmcb 

> ... if users need a version of cobra that works with older versions of go, they can use the v1.0.0 tag

The issue I have with doing this is that it seems like we are considering deprecating Go versions _just_ to satisfy this linter. We could say "If you want backward compat with older Go versions use v1.0.0. But then that cuts out those users of benefiting from some of the features that have been added since.

> Further, I would question why projects are being maintained with a version of Go that is technically EOL

You'd be surprised. :D 

IMO if we want to deprecate older Go versions that sounds good to me. I just want to make sure we prep and do it properly as to not take users by surprise. :)

Reply:
> cobra 1.0.0 is feature complete and currently backwards compatible with earlier versions of go

In fact, I thought that releasing v1.0.0 was a priority in order to then allow merging small but justified breaking changes. I've been rebasing this PR and #841 (although not strictly breaking) for +1y. I don't mind keeping doing it until the decision to accept those changes is taken.

Reply:
> IMO if we want to deprecate older Go versions that sounds good to me. I just want to make sure we prep and do it properly as to not take users by surprise. :)

Shall we open an issue to discuss it? Or will you maintainers handle it?